### PR TITLE
Add bulk memory opcode definitions

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit starts to add support in wabt's various tools for the
upcoming [bulk memory proposal][1]. This is based off the current
proposal's overview, although these may get tweaked over time!

This is also the first time I've significantly contributed to wabt, and
what I thought would be a relatively simple addition ended up being much
larger than I imagined! I didn't add many negative tests yet but if more
tests are desired please let me know!

[1]: https://github.com/webassembly/bulk-memory-operations

### CODE REVIEW

File: src/wast-parser.cc [hunk 1]
```diff
@@ -794,7 +801,12 @@ Result WastParser::ParseDataModuleField(Module* module) {
   auto field = MakeUnique<DataSegmentModuleField>(loc);
   EXPECT(Data);
   ParseVarOpt(&field->data_segment.memory_var, Var(0, loc));
```
Comment (binji) [inline src/wast-parser.cc]: `passive` should be optionally parsed first, and if set the memory index should not be parsed.

See the spec [here](https://github.com/WebAssembly/bulk-memory-operations/blob/master/document/core/text/modules.rst#data-segments) (though it's kinda hard to read)

File: src/binary-reader.cc [hunk 2]
```diff
@@ -1322,6 +1322,62 @@ Result BinaryReader::ReadFunctionBody(Offset end_offset) {
...
+          CALLBACK(OnTableInitExpr, segment);
+        }
+        break;
+      }
+
+      case Opcode::MemoryDrop:
+      case Opcode::TableDrop: {
+        ERROR_UNLESS_OPCODE_ENABLED(opcode);
+        uint8_t reserved;
+        CHECK_RESULT(ReadU8(&reserved, "reserved memory index"));
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/binary-reader.cc]: Oh oops, I actually had this write and then typo'd the serialization to emit a memory index, and then when tests failed I blindly applied the changes back here... Good catch!

**Comment by alexcrichton** [comment]
Ok, updated!

File: src/validator.cc [hunk 3]
```diff
@@ -685,20 +703,69 @@ Result Validator::EndLoopExpr(LoopExpr* expr) {
   return Result::Ok;
 }
 
+Result Validator::OnMemoryCopyExpr(MemoryCopyExpr* expr) {
+  expr_loc_ = &expr->loc;
+  CheckHasMemory(&expr->loc, Opcode::MemoryCopy);
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/validator.cc]: Want me to file a follow-up issue for this?

[the remark this answers is not in the document]
Comment (binji) [reply src/validator.cc]: Sure, thanks!

[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/validator.cc]: https://github.com/WebAssembly/wabt/issues/928

**Pull request merged**